[SPO] Add logic to handle sitegroups in ACL for drive_items and sites

jedrazb opened the pull request:
## Closes https://github.com/elastic/enterprise-search-team/issues/5837

This change ensures that site groups are captured when constructing ACLs for drive_items and sites.

The logic mirrors how the `/roleassignments` endpoint provides permissions for `site_pages`, `site_lists`, and `site_list_items`. This endpoint returns site groups alongside their associated users in a single response. As a result, we don't have to "manage" site groups with non-global IDs; instead, we can directly add associated users to the ACL.

Therefore, when fetching data from the `/permissions` endpoint using the GraphAPI, we've introduced logic to handle site groups. For each site group retrieved from the permissions endpoint or the user information list (for site access), we issue an API call to gather associated users for a given site within the context of a site collection. We then directly populate these user emails into the ACL, much like we do with `/roleassignments`.

This update eliminates the need for any additional steps in the user access control sync.

## Checklists

<!--You can remove unrelated items from checklists below and/or add new
items that may help during the review.-->

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1280,6 +1289,21 @@ def _is_site_admin(user):
+                    site["webUrl"], site_group_id
+                ):
+                    site_group_user_name = site_group_user.get(
+                        "UserPrincipalName", None
```
super nit, but `None` is already the default value for a `get()` where the key doesn't exist, so you don't need to provide it as an explicit arg.

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1732,13 +1768,32 @@ def _get_id(permissions, identity):
+                access_control.append(_prefix_email(site_user_email))
+
+            if site_group_id:
+                async for site_group_user in self.client.site_groups_users(
+                    site_web_url, site_group_id
+                ):
```
IIUC, this will re-fetch the memberships of a given site group for every single drive item that is shared with that given site group.

This could be pretty inefficient and use up a lot of API quota. Would it be valuable to first check an in-memory cache to see if the members have already been looked up for this group, and if so, grab them from the cache, and if not, fetch them and add them to the cache?

jedrazb replied:
That's a great idea to save the API quota. One potential issue (apart from mem footprint) is the case connector keeps running for a long time. Whenever `get_docs` or `get_docs_incrementally` we need to invalidate the cache to fetch up-to-date group members

jedrazb replied:
You are right! For some reason I thought that the data source instance would be kept in memory inbetween the syncs. In that case we don't need to purge the cache. I will amend the PR

seanstory reviewed (commented):
one nit, one more serious comment. But great work and turnaround time 👏 

Might be worth adding a unit test to actually validate this functionality - maybe add back the specs that were removed when we originally took out site_group and site_user, too?

jedrazb commented on connectors/sources/sharepoint_online.py:
```diff
@@ -564,13 +564,18 @@ async def user_information_list(self, site_id):
+                    yield site_group_user
         except NotFound:
-            return {}
+            return
```
Sure, I can add a warning log to be extra safe here. I followed a pattern of handling `NotFound` exceptions from other client methods.

jedrazb commented:
Addressed review feedback:
- added solid unit tests
- added in-mem cache with purge policy (triggered whenever a full/incremental sync is triggered)

jedrazb commented:
> could you please calculate the approximate memory footprint? 

@vidok Approximately the mem-footprint would be ~ `O((# of sitegroups) x (avg number of members in a sitegroup))`

Note that sitegroup is a different from Team365 group or dynamic group... from my observation for each site we have at least 3 default sitegroups (owners, visitors and members) and maybe more, each of that group has members. 

The actual footprint depends on the sharepoint setup of a client. In our test instance we had a few sites and few groups with a few members so it was tiny cache.

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1426,7 +1452,36 @@ async def process_user(user):
+        """
+
+        # Check cache first
+        if site_group_id in self.site_group_cache:
```
is this `site_group_id` one that's globally unique? It looks like it's just coming off of `user.get("id")`, which I thought would be values like `1`, `2`, etc.

jedrazb replied:
I somehow wanted to combine `site_web_url, site_group_id` into globally unique id for caching but I totally forgot to do it...

jedrazb replied:
Should be resolved now

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1426,7 +1452,36 @@ async def process_user(user):
+
+        # Check cache first
+        if site_group_id in self.site_group_cache:
+            return self.site_group_cache[site_group_id]
```
let's add some debug logging for if we find the group in the cache

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1732,13 +1797,29 @@ def _get_id(permissions, identity):
 
             user_id = _get_id(granted_to_v2, "user")
             group_id = _get_id(granted_to_v2, "group")
+            site_group_id = _get_id(granted_to_v2, "siteGroup")
+            site_user_email = _get_email(granted_to_v2, "siteUser")
```
Does anything add items to the user ALC list with `siteGroup` and `siteUser` prefixes? It looks like the logic gated under
```
if is_sharepoint_group(user):
```
1. doesn't add a group permission/id at all
2. doesn't prefix the user ids with `siteUser`

seanstory commented on connectors/sources/sharepoint_online.py:
```diff
@@ -1464,9 +1464,14 @@ async def site_group_users(self, site_web_url, site_group_id):
         - list: List of users for the given site group.
         """
 
+        cache_key = (site_web_url, site_group_id)
```
TIL you can key a python dict with a tuple!

jedrazb commented:
backport --pr 1659 --autoMerge --autoMergeMethod squash

jedrazb commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|pre-8.10-stable|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/connectors-python/1669">](https://github.com/elastic/connectors-python/pull/1669)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.3.0 -->

jedrazb commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|pre-8.10-stable|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/connectors-python/1670">](https://github.com/elastic/connectors-python/pull/1670)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.3.0 -->

jedrazb commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|pre-8.10-stable|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/connectors-python/1671">](https://github.com/elastic/connectors-python/pull/1671)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.3.0 -->

jedrazb commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|pre-8.10-stable|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/connectors-python/1672">](https://github.com/elastic/connectors-python/pull/1672)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.3.0 -->

seanstory commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.10|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/connectors-python/1712">](https://github.com/elastic/connectors-python/pull/1712)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
